src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoCommon.cs

```diff
@@ -119,72 +119,88 @@ public override Type GetConcreteType(Type parentType)
-                return instanceOfQueue;
+            }
+            catch (Exception ex)
+            {
+                exception = ex;
             }
 
             // TODO: Use reflection to support types implementing Stack or Queue.
 
-            ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(collectionPropertyInfo.DeclaredPropertyType, jsonPath);
+            ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(collectionPropertyInfo.DeclaredPropertyType, jsonPath, exception);
             return null;
         }
 
         public override object CreateDerivedDictionaryInstance(JsonPropertyInfo collectionPropertyInfo, IDictionary sourceDictionary, string jsonPath, JsonSerializerOptions options)
         {
+            Exception exception = null;
             object instance = collectionPropertyInfo.DeclaredTypeClassInfo.CreateObject();
 
-            if (instance is IDictionary instanceOfIDictionary)
+            try
             {
-                foreach (DictionaryEntry entry in sourceDictionary)
+                if (instance is IDictionary instanceOfIDictionary && !instanceOfIDictionary.IsReadOnly)
                 {
-                    instanceOfIDictionary.Add((string)entry.Key, entry.Value);
+                    foreach (DictionaryEntry entry in sourceDictionary)
+                    {
+                        instanceOfIDictionary.Add((string)entry.Key, entry.Value);
+                    }
+                    return instanceOfIDictionary;
                 }
-                return instanceOfIDictionary;
-            }
-            else if (instance is IDictionary<string, TRuntimeProperty> instanceOfGenericIDictionary)
-            {
-                foreach (DictionaryEntry entry in sourceDictionary)
+                else if (instance is IDictionary<string, TRuntimeProperty> instanceOfGenericIDictionary && !instanceOfGenericIDictionary.IsReadOnly)
                 {
-                    instanceOfGenericIDictionary.Add((string)entry.Key, (TRuntimeProperty)entry.Value);
+                    foreach (DictionaryEntry entry in sourceDictionary)
+                    {
+                        instanceOfGenericIDictionary.Add((string)entry.Key, (TRuntimeProperty)entry.Value);
+                    }
+                    return instanceOfGenericIDictionary;
                 }
-                return instanceOfGenericIDictionary;
+            }
+            catch (Exception ex)
+            {
+                exception = ex;
             }
 
             // TODO: Use reflection to support types implementing SortedList and maybe immutable dictionaries.
 
             // Types implementing SortedList and immutable dictionaries will fail here.
-            ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(collectionPropertyInfo.DeclaredPropertyType, jsonPath);
+            ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(collectionPropertyInfo.DeclaredPropertyType, jsonPath, exception);
```

Comment:
We should throw a `NotSupportedException` here using https://github.com/dotnet/corefx/blob/6f8983d64c67fa86ac6f0338cde08d82e51cc805/src/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs#L21.